contributors/design-proposals/deployment.md

```diff
@@ -159,56 +165,89 @@ https://github.com/kubernetes/kubernetes/issues/11234 tracks updating kubelet
 and https://github.com/kubernetes/kubernetes/issues/12615 tracks adding
 LastTransitionTime to PodCondition.
 
+### TemplateGeneration
+
+Hashing an API object such as PodTemplateSpec means that the resulting hash is subject to
+change due to API changes in PodTemplateSpec between minor versions of Kubernetes when new
+API changes are introduced. A new API field will be introduced in DeploymentSpec called
+TemplateGeneration. The new field will be a counter that will track the number of
+times an update happened in the PodTemplateSpec of the Deployment, similarly to how
+metadata.Generation tracks updates in the Spec of all first-class API objects. Unlike
+metadata.Generation, this field can be initialized by users in order to facilitate adoption
+of existing ReplicaSets. This is similar to how it is already used by DaemonSets.
+
+TemplateGeneration is not authoritative and only helps in constructing the name for the new
+ReplicaSet in case there is no other ReplicaSet that matches the Deployment. The Deployment
+controller still decides what is the new ReplicaSet by comparing PodTemplateSpecs.
+If no matching ReplicaSet is found, the controller will try to create a new ReplicaSet using
+its current TemplateGeneration.
```

Comment:
TemplateGeneration can't be used for migrating old ReplicaSets. This paragraph talks about what will happen to new ReplicaSets if we switch to TemplateGeneration. I will try to make it clearer.